docs/source/llms/notebooks/chat-model-tool-calling.ipynb

```diff
@@ -0,0 +1,501 @@
+    "2. Dataclasses for expected inputs / outputs for a better development experience\n",
+    "\n",
+    "For a more in-depth exploration of ChatModel, please check out the [detailed guide](https://mlflow.org/docs/latest/llms/chat-model-guide/index.html).\n",
+    "\n",
+    "In this tutorial, we'll be building a simple OpenAI wrapper that makes use of the tool calling support (released in MLflow 2.17.0)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Environment setup\n",
+    "\n",
+    "First, let's set up the environment. We'll need the OpenAI Python SDK, as well as MLflow >= 2.17.0. We'll also need to set our OpenAI API key in order to use the SDK."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Note: you may need to restart the kernel to use updated packages.\n"
+     ]
+    }
+   ],
+   "source": [
+    "%pip install 'mlflow>=2.17.0' 'openai>=1.0' -qq"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "from getpass import getpass\n",
+    "\n",
+    "os.environ[\"OPENAI_API_KEY\"] = getpass(\"Enter your OpenAI API key: \")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Launch the MLflow Tracking Server\n",
+    "\n",
+    "Next, we'll need to launch the [MLflow Tracking server](https://mlflow.org/docs/latest/tracking/server.html). This is an HTTP server that serves multiple REST endpoints that we'll be using in this tutorial.\n",
+    "\n",
+    "In a separate terminal shell, run `mlflow server`. By default, this will launch the tracking server on `http://localhost:5000`."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Step 1: Creating the tool definition\n",
+    "\n",
+    "Since our model will be a tool calling model, we'll be defining a `get_weather` tool that we can pass to the OpenAI model to use. We do this by using [`mlflow.types.llm.FunctionToolDefinition`](https://mlflow.org/docs/latest/python_api/mlflow.types.html#mlflow.types.llm.FunctionToolDefinition) to describe the parameters that our tool accepts. The format of this dataclass is aligned with the OpenAI spec:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from mlflow.types.llm import (\n",
+    "    FunctionToolDefinition,\n",
+    "    ParamProperty,\n",
+    "    ToolParamsSchema,\n",
+    ")\n",
+    "\n",
+    "# a sample tool definition. we use the `FunctionToolDefinition`\n",
+    "# class to describe the name and expected params for the tool.\n",
+    "# for this example, we're defining a simple tool that returns\n",
+    "# the weather for a given city.\n",
+    "weather_tool = FunctionToolDefinition(\n",
```

Comment:
Unrelated to this PR: What do you think about the idea of inferring this from a single function? I think all of these settings are present in the function definition + type hint. Sth like this:

```
ToolDefinition.from_function(get_weather)
```

Reply:
yeah that would be a great feature! i'll add it to the backlog for chatmodel improvements